internal/coreinternal/scraperinttest/scraperint.go

```diff
@@ -50,38 +51,49 @@ func NewIntegrationTest(f receiver.Factory, cr testcontainers.ContainerRequest,
 }
 
 type IntegrationTest struct {
-	containerRequest testcontainers.ContainerRequest
+	containerRequest *testcontainers.ContainerRequest
 
 	factory      receiver.Factory
 	customConfig customConfigFunc
 
 	expectedFile   string
 	compareOptions []pmetrictest.CompareMetricsOption
 	compareTimeout time.Duration
+
+	dumpActualOnFailure bool
 }
 
 func (it *IntegrationTest) Run(t *testing.T) {
 	ctx := context.Background()
-	require.NoError(t, it.containerRequest.Validate())
-
-	var container testcontainers.Container
-	var err error
-	require.Eventually(t, func() bool {
-		container, err = testcontainers.GenericContainer(ctx,
-			testcontainers.GenericContainerRequest{
-				ContainerRequest: it.containerRequest,
-				Started:          true,
-			})
-		return err == nil
-	}, 5*time.Minute, time.Second)
-	defer func() {
-		require.NoError(t, container.Terminate(ctx))
-	}()
 
-	cfg := it.factory.CreateDefaultConfig()
-	host, mappedPort := containerInfo(ctx, t, container)
-	it.customConfig(cfg, host, mappedPort)
+	var ci *ContainerInfo
+	if it.containerRequest != nil {
+		require.NoError(t, it.containerRequest.Validate())
+		for _, port := range it.containerRequest.ExposedPorts {
+			require.False(t, strings.ContainsRune(port, ':'), errExposedPort)
+		}
+
+		var container testcontainers.Container
+		var err error
+		require.Eventually(t, func() bool {
```

Comment:
In a follow up PR, would it possible to add `t.Log` on error?

I've found that trying to chase up `Eventually` conditions and trying to understand what is happening is difficult.

Reply:
👍 This is on my list. I think it'll make sense to extract a common mechanism for logging the last failure from `Eventually` loops.